src/shared/sol-conffile.c

```diff
@@ -58,87 +60,207 @@ sol_conffile_entry_sort_cb(const void *data1, const void *data2)
     return strcasecmp(e1->id, e2->id);
 }
 
-static struct sol_ptr_vector
-sol_conffile_keyfile_to_vector(GKeyFile *keyfile)
+static void
+sol_free_conffile_entry(struct sol_conffile_entry *entry)
 {
-    struct sol_ptr_vector pv;
-    gchar **groups, **groups_ptr;
-    struct sol_conffile_entry *entry = NULL;
+    free(entry->id);
+    free(entry->type);
+    if (entry->options) {
+        char **tmp = entry->options;
+        while (*tmp) {
+            free(*tmp);
+            tmp++;
+        }
+        free(entry->options);
+    }
+    free(entry);
+}
 
-    sol_ptr_vector_init(&pv);
+static int
+sol_conffile_set_entry_options(struct sol_conffile_entry *entry, struct sol_json_token options_object)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    struct sol_ptr_vector vec_options;
+    enum sol_json_loop_reason reason;
+
+    sol_ptr_vector_init(&vec_options);
+
+    sol_json_scanner_init_from_token(&scanner, &options_object);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        int key_len, value_len;
+        char *tmp;
+        key_len = sol_json_token_get_size(&key) - 2; // remove quotes from the key.
+        value_len = sol_json_token_get_size(&value); // do not remove quotes from the value. (not everything is a string here)
+        if (asprintf(&tmp, "%.*s=%.*s", key_len, key.start + 1, value_len, value.start) <= 0) {
```

Comment:
Might be a good idea to check if `key_len` is zero before calling `asprintf()` and abort the loading procedure.